### PR TITLE
feat: add three lines chat list item and mention notifications

### PR DESCRIPTION
**Pull request by imGok**
This is related to #100 
- Added Three lines display if chat item is a group
- Added Dot animations when user is typing
- Adding Mention badge at the end of chat list item 

demo_three_lines

### CODE REVIEW

File: lib/pages/chat_list/chat_list_item.dart [hunk 1]
```diff
@@ -119,8 +119,7 @@ class ChatListItem extends StatelessWidget {
   Widget build(BuildContext context) {
     final isMuted = room.pushRuleState != PushRuleState.notify;
     final typingText = room.getLocalizedTypingText(context);
-    final ownMessage =
-        room.lastEvent?.senderId == Matrix.of(context).client.userID;
+    final ownMessage = room.lastEvent?.senderId == Matrix.of(context).client.userID;
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:122]: IMO, you should change the name of variable for clear meaning. `ownLastMessage = room.lastEvent?.senderId == Matrix.of(context).client.userID;`

File: lib/pages/chat_list/chat_list_item.dart [hunk 2]
```diff
@@ -130,6 +129,7 @@ class ChatListItem extends StatelessWidget {
     final displayname = room.getLocalizedDisplayname(
       MatrixLocals(L10n.of(context)!),
     );
+    final isGroup = room.isDirectChat == false;
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:132]: `final isGroup = !room.isDirectChat;`

File: lib/pages/chat_list/chat_list_item.dart [hunk 3]
```diff
@@ -143,42 +143,46 @@ class ChatListItem extends StatelessWidget {
...
-                            style: Theme.of(context).textTheme.titleMedium
-                                ?.merge(const TextStyle(
-                                  overflow: TextOverflow.ellipsis, 
-                                  letterSpacing: 0.15)),
+                            style: Theme.of(context).textTheme.titleMedium?.merge(
+                                  TextStyle(
+                                    overflow: TextOverflow.ellipsis,
+                                    letterSpacing: 0.15,
+                                    color:
+                                        unread ? const Color(0xFF1C1B1F) : const Color(0xFF787579),
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:183]: You should use Theme.of to look up color as much as possible, don't use raw Color, in this case please use: `Theme.of(context).colorScheme.onSurface`

File: lib/pages/chat_list/chat_list_item.dart [hunk 4]
```diff
@@ -201,140 +205,189 @@ class ChatListItem extends StatelessWidget {
...
-                  if (typingText.isEmpty &&
-                      ownMessage &&
-                      room.lastEvent!.status.isSending) ...[
-                    const SizedBox(
-                      width: 16,
-                      height: 16,
-                      child: CircularProgressIndicator.adaptive(strokeWidth: 2),
-                    ),
-                    const SizedBox(width: 4),
+                    isGroup
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:208]: IMO, i think if code is nested too much, you should separate it to a helper function, or a widget

File: lib/pages/chat_list/chat_list_item.dart [hunk 5]
```diff
@@ -201,140 +205,189 @@ class ChatListItem extends StatelessWidget {
...
+                                    snapshot.data!.displayName!,
+                                    overflow: TextOverflow.ellipsis,
+                                    maxLines: 1,
+                                    softWrap: false,
+                                    style: Theme.of(context).textTheme.titleMedium?.merge(
+                                          TextStyle(
+                                            overflow: TextOverflow.ellipsis,
+                                            letterSpacing: 0.15,
+                                            color: unread
+                                                ? const Color(0xFF1C1B1F)
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:225]: please change color

File: lib/pages/chat_list/chat_list_item.dart [hunk 6]
```diff
@@ -201,140 +205,189 @@ class ChatListItem extends StatelessWidget {
...
-                                            .onPrimaryContainer,
-                                fontSize: 13,
-                              ),
-                            )
+                                  '@',
+                                  style: TextStyle(
+                                    color: isMentionned
+                                        ? Theme.of(context).colorScheme.onPrimary
+                                        : Theme.of(context).colorScheme.onPrimaryContainer,
+                                    fontSize: 12,
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:347]: Screenshot 2023-05-26 at 14 10 34

`Theme.of(context).textTheme.labelMedium`

File: lib/pages/chat_list/chat_list_item.dart [hunk 7]
```diff
@@ -201,140 +205,189 @@ class ChatListItem extends StatelessWidget {
...
-          ],
+              const SizedBox(width: 4),
+              AnimatedContainer(
+                duration: FluffyThemes.animationDuration,
+                curve: FluffyThemes.animationCurve,
+                padding: const EdgeInsets.symmetric(horizontal: 7),
+                height: unreadBubbleSize,
+                width: room.notificationCount == 0 && !unread && !room.hasNewMessages
+                    ? 0
+                    : (unreadBubbleSize - 9) * room.notificationCount.toString().length + 9,
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:363]: What is 9 ? IMO, you should make a const for this number

Comment (hoangdat) [reply lib/pages/chat_list/chat_list_item.dart:363]: dont use the magic number

File: lib/pages/chat_list/chat_list_item.dart (hunk 2 above)
Comment (imGok) [reply lib/pages/chat_list/chat_list_item.dart]: I was a bit tired I guess 😭

File: lib/utils/string_extension.dart [hunk 8]
```diff
@@ -47,4 +47,30 @@ extension StringCasingExtension on String {
   String toTomMatrixId() {
     return '@$this:tom-dev.xyz';
   }
+
+  List<String> getMentionsFromMessage() {
+    final RegExp regex = RegExp(r"@\[([^\]]+)\]");
+    final Iterable<Match> matches = regex.allMatches(this);
+
+    final List<String> mentions = [];
+    for (final Match match in matches) {
```
Comment (sherlockvn) [inline lib/utils/string_extension.dart:56]: IMO, you can make code easier to read by using `matches.where((match) => match.group(0) != null)`

File: lib/utils/string_extension.dart [hunk 9]
```diff
@@ -47,4 +47,30 @@ extension StringCasingExtension on String {
...
+      if (mention != null) {
+        mentions.add(mention);
+      }
+    }
+    return mentions;
+  }
+
+  List<String> getAllMentionedUserIdsFromMessage(Room room) {
+    final List<String> mentionUserIds = [];
+    for (final String mention in getMentionsFromMessage()) {
```
Comment (sherlockvn) [inline lib/utils/string_extension.dart:67]: also use .where like above

File: lib/utils/string_extension.dart [hunk 10]
```diff
@@ -47,4 +47,30 @@ extension StringCasingExtension on String {
...
+      final String? userId = room.lastEvent!.room.getMention(mention);
```
Comment (sherlockvn) [inline lib/utils/string_extension.dart:68]: When i firstly create the room, does the room have any event yet

**Comment by hoangdat** [comment]
- How about other requirement of your task? 
- What you want to deliver in this task?
- What task you are working for? No reference?

**Comment by hoangdat** [comment]
it is not acceptable for my site.

**Comment by hoangdat** [comment]
It take me and the team a lot of time to review and follow what you want to deliver, please safe the time of the people

File: lib/pages/chat_list/chat_list_item.dart [hunk 11]
```diff
@@ -201,142 +202,215 @@ class ChatListItem extends StatelessWidget {
...
+
+  Row typingTextWidget(String typingText, BuildContext context) {
+    return Row(
+      crossAxisAlignment: CrossAxisAlignment.center,
+      mainAxisAlignment: MainAxisAlignment.start,
+      children: [
+        Flexible(
+          child: Text(
+              typingText,
+              style: Theme.of(context).textTheme.titleMedium?.merge(
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list_item.dart:366]: Should be the `label/large`.

File: lib/pages/chat_list/chat_list_item_style.dart [hunk 12]
```diff
@@ -5,4 +5,26 @@ class ChatListItemStyle {
   static Color? get readIconColor => LinagoraRefColors.material().tertiary[20];
   static double get readIconSize => 16;
   static double get editIconSize => 14;
-}
\ No newline at end of file
+  static double get mentionIconWidth => 20;
+  static Color get readMessageColor => const Color(0xFF787579);
```
[the remark this answers is not in the document]
Comment (imGok) [reply lib/pages/chat_list/chat_list_item_style.dart:9]: It's in lastSenderWidget 
`...
                            color: unread
                                ? Theme.of(context).colorScheme.onSurfaceVariant
                                : ChatListItemStyle.readMessageColor,`

File: lib/utils/string_extension.dart (hunk 8 above)
Comment (imGok) [reply lib/utils/string_extension.dart:56]: I don't want to have duplicate mentions

File: lib/utils/string_extension.dart (hunk 9 above)
Comment (imGok) [reply lib/utils/string_extension.dart:67]: in this case ok

File: lib/utils/string_extension.dart (hunk 10 above)
Comment (imGok) [reply lib/utils/string_extension.dart:68]: I'm not sure, I'll use nullable "?" instead

**Comment by imGok** [comment]
@hoangdat all good @sherlockvn

File: lib/utils/string_extension.dart [hunk 13]
```diff
@@ -2,14 +2,12 @@ import 'package:fluffychat/widgets/matrix.dart';
...
+import 'package:matrix/matrix.dart';
 
 extension StringCasingExtension on String {
   String removeDiacritics() {
-    const withDia =
-        'ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž';
-    const withoutDia =
-        'AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz';
+    const withDia = 'ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž';
+    const withoutDia = 'AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz';
```
Comment (hoangdat) [inline lib/utils/string_extension.dart:10]: what is it?

File: test/string_extension_test.dart [hunk 14]
```diff
@@ -0,0 +1,34 @@
+import 'package:fluffychat/utils/string_extension.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
```
Comment (hoangdat) [inline test/string_extension_test.dart:4]: How about the case : `message include @ in content`?

File: lib/widgets/twake_components/twake_loading/flashing_circle_widget.dart [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      animation: repeatingController,
+      builder: (context, child) {
+        final circleFlashPercent = dotIntervals[index].transform(
+          repeatingController.value,
+        );
+        final circleColorPercent = sin(pi * circleFlashPercent);
+
+        return Container(
+          width: 8 * repeatingController.value,
+          height: 8 * repeatingController.value,
```
Comment (hoangdat) [inline lib/widgets/twake_components/twake_loading/flashing_circle_widget.dart:33]: what is 8?

File: lib/pages/chat_list/chat_list_item.dart [hunk 16]
```diff
@@ -119,17 +121,14 @@ class ChatListItem extends StatelessWidget {
...
-        room.lastEvent?.senderId == Matrix.of(context).client.userID;
+
+    final ownLastMessage = room.lastEvent?.senderId == Matrix.of(context).client.userID;
     final unread = room.isUnread || room.membership == Membership.invite;
-    final unreadBubbleSize = unread || room.hasNewMessages
-        ? room.notificationCount > 0
-            ? 20.0
-            : 14.0
-        : 0.0;
+    final unreadBubbleSize = ChatListItemStyle.unreadBadgeSize(unread, room.hasNewMessages, room.notificationCount > 0);
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list_item.dart:127]: what is bubble here?

File: lib/pages/chat_list/chat_list_item.dart [hunk 17]
```diff
@@ -201,142 +202,215 @@ class ChatListItem extends StatelessWidget {
...
+              FutureBuilder<User?>(
+                future: room.lastEvent?.fetchSenderUser(),
+                builder: (context, snapshot) {
+                  if (snapshot.data == null) return const SizedBox.shrink();
+                  return Text(
+                    snapshot.data!.displayName!,
+                    overflow: TextOverflow.ellipsis,
+                    maxLines: 1,
+                    softWrap: false,
+                    style: Theme.of(context).textTheme.titleMedium?.merge(
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list_item.dart:398]: IMO, it is label large

image

File: lib/widgets/twake_components/twake_loading/flashing_circle_widget.dart (hunk 15 above)
Comment (imGok) [reply lib/widgets/twake_components/twake_loading/flashing_circle_widget.dart:33]: 8 is the size of 3 "bubbles" in loading animation

File: lib/utils/string_extension.dart (hunk 13 above)
Comment (imGok) [reply lib/utils/string_extension.dart:10]: It's my auto indentation, I'll revert it

File: test/string_extension_test.dart (hunk 14 above)
Comment (hoangdat) [reply test/string_extension_test.dart:4]: and `message has @[] in content`?

File: lib/utils/string_extension.dart (hunk 13 above)
Comment (hoangdat) [reply lib/utils/string_extension.dart:10]: can not understand

**Review by hoangdat — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**